### PR TITLE
Connection: extracting the Jetpack's authorization webhook

### LINKED ISSUE #15932: Connection Package: Blank page when retrying authorization
If the `authorization` doesn't succeed the first time in Calypso, and the user clicks `Try again`, they'll be redirected to a blank page.

#### Steps to reproduce the issue
- Make sure Jetpack-the-plugin is **not** installed in your site.
- Go to the `Client Example` plugin, start the Calypso flow for authorization.
- Before clicking "Approve" in Calypso, add something to your local site that would make the request fail. I just killed my `ngrok` tunnel.
- Click `Approve`, you'll get an error notice with a "Try again" link.
- Take your local site back online.
- Click `Try Again`. You'll be redirected to a blank page.

The problem is that the page you're redirected is `https://YOUR_SITE/wp-admin/admin.php?action=authorize&_wpnonce=8ab6029e62&redirect=<REDIRECT_URL>`, and the logic to process that is in Jetpack-the-plugin: https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/class.jetpack.php#L4186-L4195

#### What I expected
I expected the retry flow to work the same way as in Jetpack.

#### What happened instead
Blank page.

### PR DESCRIPTION
**Pull request by sergeymitr**
Fixes #15932.

The `authorize` webhook is used as a fallback action for Calypso authorize.
This PR is moving this webhook from Jetpack to `jetpack-connection`, making it available for plugins using the package without the Jetpack plugin.

#### Changes proposed in this Pull Request:
This commit introduces following changes:
- The `authorize` webhook is moved from Jetpack plugin into the Connection package.
- New class `Webhooks` responsible for handling the `authorize` and other requests we will move from Jetpack or create from scratch.
- Refactoring of the `Jetpack_Client_Server` class, deprecating some of its methods.
- Adjusting the existing `authorize` action handler to make it use the new `Webhooks` class, and deprecating it in favor of the new webhook.
- Unit tests for the new `Webhooks` class.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Switch Jetpack Beta to the branch, then disable Jetpack.
2. Install Client Example, run its `composer install` if needed. Delete its `vendor` directory, and symlink the one from `jetpack-dev`:
```bash
cd wp-content/plugins/client-example
composer install
rm -rf vendor
ln -s ../jetpack-dev/vendor
```
3. Go to "Client Example -> Simple UI with Calypso", click "Set Up Connection", but **do not** click "Approve".
4. Open `wp-config.php`, add `die();` into the beginning of the file, and click "Approve". You should see the `We had trouble connecting.` error message, and the "Try again" link.
5. Remove the `die();` from the file, and click on the "Try again" link. You should get redirected to the WP-Admin Dashboard. Go to "Client Example" and confirm that you're fully connected (both "Blog token" and "user token sections indicate that you're connected).
6. Activate the Jetpack plugin, disconnect it.
7. Go to "Settings -> Jetpack Constants" and enable the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant.
8. Go to Jetpack Dashboard and connect Jetpack. After going through the Calypso connection flow, you should get redirected to Jetpack Dashboard with Jetpack fully connected.
9. Enable the `WP_DEBUG` constant in `wp-config.php`.
10. Load the URL `wp-admin/admin.php?page=jetpack&action=authorize`. You should see the `webhook is deprecated` error notice.
11. Disable the `WP_DEBUG` constant and reload that URL. You should get redirected to Jetpack Dashboard with the message "Your Jetpack is already connected."

#### Proposed changelog entry for your changes:
_Not sure if needed, but anyway:_
- Resolved an issue with connecting plugins to WordPress.com.

### CODE REVIEW

File: projects/packages/connection/src/class-webhooks.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	 * Perform the authorization action.
+	 */
+	public function handle_authorize() {
+		if ( $this->connection->is_active() && $this->connection->is_user_connected() ) {
+			$redirect_url = apply_filters( 'jetpack_client_authorize_already_authorized_url', admin_url() );
+			wp_safe_redirect( $redirect_url );
+
+			return;
+		}
+		do_action( 'jetpack_client_authore_processing' );
```
Comment (kbrown9) [inline projects/packages/connection/src/class-webhooks.php:77]: I think this is a typo. It should probably be:

```suggestion
		do_action( 'jetpack_client_authorize_processing' );
```

File: projects/packages/connection/src/class-webhooks.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	public static function init( $connection ) {
+		$webhooks = new static( $connection );
+
+		add_action( 'init', array( $webhooks, 'controller' ) );
+	}
+
+	/**
+	 * The "controller" decides which handler we need to run.
+	 */
+	public function controller() {
```
Comment (kbrown9) [inline projects/packages/connection/src/class-webhooks.php:49]: Do you think unit tests for `Webhooks::controller()` would provide any value?

**Review by kbrown9 — DISMISSED** [review]
I tested using the provided testing instructions, and everything functioned as expected with Client Example.

I wasn't really able to test much with Jetpack. For example, step 8 (_Go to Jetpack Dashboard and connect Jetpack. You should get redirected to Jetpack Dashboard with Jetpack fully connected.)_ didn't happen in my tests. I was just redirected to Calypso.

I left a couple of small comments inline.

File: projects/packages/connection/src/class-webhooks.php (hunk 1 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-webhooks.php]: Good catch, thanks Kim!
I added that to unit tests, as I should've done earlier.

File: projects/packages/connection/src/class-webhooks.php (hunk 2 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-webhooks.php:49]: Completed, thanks for pointing that out.

**Comment by sergeymitr** [comment]
Hi @kbrown9.

Thanks for your review.

> I wasn't really able to test much with Jetpack. For example, step 8 (Go to Jetpack Dashboard and connect Jetpack. You should get redirected to Jetpack Dashboard with Jetpack fully connected.) didn't happen in my tests. I was just redirected to Calypso.

I should've been more clear, sorry.
What I meant is being redirected to Jetpack Dashboard after going through the regular Calypso flow. I edited the testing instructions to reflect that.

I also implemented the changes you suggested, please re-review when you get a few minutes.

**Review by kbrown9 — APPROVED** [review]
This looks good to me, and works well in my tests with client-example.

**Pull request merged**